### PR TITLE
[DEVHAS-262] cdq outerloop behavior improvement

### PR DESCRIPTION
**Pull request by yangcao77**
### What does this PR do?:

This PR improves the cdq behavior for better outerloop support. 
CDQ now validates the devfile content and the behavior will be:

    devfile parser error - if a devfile has invalid schema or there is an error parsing it, CDQ should return the error
    invalid kubernetes uri/invalid kubernetes definition - CDQ needs to error out in this situation because an outerloop file has been provided
    missing devfile kubernetes/image components - this means that the devfile contains no outerloop definition and CDQ ignores the devfile from provided repo, and needs to match a devfile from registry. Unless the `devfileURL` is specifically being set, this case, the cdq will error out as the requested devfile is invalid for outerloop. 
    missing deploy command/apply command - if devfile contains more than one kubernetes components but with no deploy command being defined, or contains more than one image components but with no apply command being defined. CDQ will error out. Since HAS will have no idea on which definition should be used when come to component creation. 

### Which issue(s)/story(ies) does this PR fixes:

https://issues.redhat.com/browse/DEVHAS-262

### PR acceptance criteria:

- [x] Unit/Functional tests

  

- [ ] Documentation 

   

- [ ] Client Impact

### CODE REVIEW

**Comment by johnmcollier** [comment]
@yangcao77 @maysunfaisal 

For this specific error scenario:
```
missing devfile kubernetes components/image component - this means that the devfile contains no outerloop definition and CDQ ignores the devfile from provided repo. It will fall back to the behavior that no devfile exist in the repo. which mean if dockerfile exist, a devfile will be generated from the dockerfile, else needs to match a devfile from registry.
```

are we sure that we want to match a devfile from the sample in this case, and override the user's devfile? Wouldn't it be better to fall back on gitops generation from within HAS?

**Comment by yangcao77** [comment]
@johnmcollier Yes, we discussed in HAS team meeting for 2-3 times, and this is the behavior we want now. which the valid devfile content will be returned by CDQ and stored in the component spec. ideally there should be a warning displayed after CDQ is done in UI, but currently HAC does not accept/display any warnings from HAS.

File: pkg/devfile/devfile.go [hunk 1]
```diff
@@ -734,3 +741,132 @@ func UpdateLocalDockerfileURItoAbsolute(devfile data.DevfileData, dockerfileURL
...
+					break
+				}
+			}
+			if !found {
+				err = fmt.Errorf("found more than one image components, but no deploy command associated with any being defined in the devfile from %s", URL)
+				log.Error(err, "failed to validate devfile")
+				return shouldIgnoreDevfile, nil, err
+			}
+		}
+		// TODO: if only one image component, should return a warning that no apply command being defined
```
Comment (johnmcollier) [inline pkg/devfile/devfile.go:868]: Does this still need to be done?

**Review by johnmcollier — COMMENTED** [review]
Codecov reports the overall coverage of HAS dropping by -1.56%, would it be possible to try to increase the coverage?

File: pkg/devfile/devfile.go [hunk 2]
```diff
@@ -734,3 +741,132 @@ func UpdateLocalDockerfileURItoAbsolute(devfile data.DevfileData, dockerfileURL
...
+					break
+				}
+			}
+			if !found {
+				err = fmt.Errorf("found more than one kubernetes components, but no deploy command associated with any being defined in the devfile from %s", URL)
+				log.Error(err, "failed to validate devfile")
+				return shouldIgnoreDevfile, nil, err
+			}
+		}
+		// TODO: if only one kube component, should return a warning that no deploy command being defined
```
[the remark this answers is not in the document]
Comment (yangcao77) [reply pkg/devfile/devfile.go:808]: Yes. Since HAC does not accept or display any warnings.

File: pkg/devfile/devfile.go (hunk 1 above)
Comment (yangcao77) [reply pkg/devfile/devfile.go:868]: same as above, we currently have no way to return warning to HAC

File: pkg/devfile/devfile.go [hunk 3]
```diff
@@ -734,3 +741,132 @@ func UpdateLocalDockerfileURItoAbsolute(devfile data.DevfileData, dockerfileURL
...
+						if devfileSrc.Path != "" {
+							// local devfile src with relative dockerfile uri
+							dockerfileURI = path.Join(path.Dir(URL), dockerfileURI)
+							err = parserUtil.ValidateFile(dockerfileURI)
+						} else {
+							// remote devfile src with relative dockerfile uri
+							var u *url.URL
+							u, err = url.Parse(URL)
+							if err != nil {
+								log.Error(err, fmt.Sprintf("failed to parse from %s", URL))
```
[the remark this answers is not in the document]
Comment (yangcao77) [reply pkg/devfile/devfile.go]: updated

File: pkg/devfile/devfile.go [hunk 4]
```diff
@@ -734,3 +741,132 @@ func UpdateLocalDockerfileURItoAbsolute(devfile data.DevfileData, dockerfileURL
...
+								return shouldIgnoreDevfile, nil, fmt.Errorf(fmt.Sprintf("failed to parse from %s", URL))
+							}
+							u.Path = path.Join(u.Path, dockerfileURI)
+							dockerfileURI = u.String()
+							_, err = util.CurlEndpoint(dockerfileURI)
+						}
+					}
+					if err != nil {
+						log.Error(err, fmt.Sprintf("failed to curl from the dockerfile URI %s, invalid image component: %s", URL, component.Name))
```
[the remark this answers is not in the document]
Comment (yangcao77) [reply pkg/devfile/devfile.go]: updated

**Comment by yangcao77** [comment]
> Codecov reports the overall coverage of HAS dropping by -1.56%, would it be possible to try to increase the coverage?

I have update the PR to be -1.26%. 
The coverage report alert some warnings that are not part of this PR, and the rest are some errors not easily to mock & test.

**Comment by yangcao77** [comment]
@johnmcollier I have addressed your review comments. Can you help review again when you have time?

**Review by johnmcollier — APPROVED** [review]

**Pull request merged**